File: src/dots/plugins/managers.py
"""
Package manager adapters for cross-platform dependency installation.
"""
import shutil
import subprocess
from abc import ABC, abstractmethod
from typing import List, Optional
from rich.console import Console


class PackageManager(ABC):
    """Abstract base class for package managers."""
    
    @property
    @abstractmethod
    def name(self) -> str:
        """Human-readable name of the package manager."""
        pass
    
    @property
    @abstractmethod
    def needs_sudo(self) -> bool:
        """Whether this package manager requires sudo for installation."""
        pass
    
    @abstractmethod
    def is_available(self) -> bool:
        """Check if this package manager is available on the system."""
        pass

    @abstractmethod
    def install_command(self, packages: List[str]) -> List[str]:
        """
        Return the command to install packages as a list of arguments.
        
        Returns:
            List of command arguments (e.g., ["pacman", "-S", "--noconfirm", "pkg1", "pkg2"])
        """
        pass

    @abstractmethod
    def update_command(self) -> List[str]:
        """Return the command to update package lists."""
        pass
    
    def is_package_available(self, package: str) -> bool:
        """
        Check if a package is available in this manager's repositories.
        
        Default implementation returns True. Subclasses can override for validation.
        """
        return True


class Pacman(PackageManager):
    @property
    def name(self) -> str:
        return "pacman"
    
    @property
    def needs_sudo(self) -> bool:
        return True
    
    def is_available(self) -> bool:
        return shutil.which("pacman") is not None

    def install_command(self, packages: List[str]) -> List[str]:
        return ["pacman", "-S", "--noconfirm"] + packages
        
    def update_command(self) -> List[str]:
        return ["pacman", "-Sy"]


class Apt(PackageManager):
    @property
    def name(self) -> str:
        return "apt"
    
    @property
    def needs_sudo(self) -> bool:
        return True
    
    def is_available(self) -> bool:
        return shutil.which("apt-get") is not None

    def install_command(self, packages: List[str]) -> List[str]:
        return ["apt-get", "install", "-y"] + packages
        
    def update_command(self) -> List[str]:
        return ["apt-get", "update"]


class Brew(PackageManager):
    @property
    def name(self) -> str:
        return "brew"
    
    @property
    def needs_sudo(self) -> bool:
        return False
    
    def is_available(self) -> bool:
        return shutil.which("brew") is not None

    def install_command(self, packages: List[str]) -> List[str]:
        return ["brew", "install"] + packages
        
    def update_command(self) -> List[str]:
        return ["brew", "update"]


def get_package_manager() -> Optional[PackageManager]:
    """
    Detect and return the available package manager.
    
    Priority: Pacman > Apt > Brew
    """
    if shutil.which("pacman"):
        return Pacman()
    if shutil.which("apt-get"):
        return Apt()
    if shutil.which("brew"):
        return Brew()
    return None

File: internal/plugins/managers.go
// Package plugins provides adapters for external tools and package managers.
package plugins

import (
	"os/exec"
	"sort"
)

// PackageManager defines the interface for system package managers.
type PackageManager interface {
	// Name returns the human-readable name (e.g., "pacman", "apt", "brew").
	Name() string
	// NeedsSudo indicates whether this manager requires sudo for installs.
	NeedsSudo() bool
	// IsAvailable checks if this manager is installed on the system.
	IsAvailable() bool
	// InstallCommand returns the command to install the given packages.
	InstallCommand(packages []string) []string
}

// Pacman adapts the Arch Linux package manager.
type Pacman struct{}

func (Pacman) Name() string           { return "pacman" }
func (Pacman) NeedsSudo() bool        { return true }
func (Pacman) IsAvailable() bool      { return exec.Command("pacman", "--version").Run() == nil }

func (Pacman) InstallCommand(packages []string) []string {
	cmd := []string{"pacman", "-S", "--noconfirm"}
	return append(cmd, packages...)
}

// Apt adapts the Debian/Ubuntu package manager.
type Apt struct{}

func (Apt) Name() string      { return "apt" }
func (Apt) NeedsSudo() bool   { return true }
func (Apt) IsAvailable() bool { return exec.Command("apt-get", "--version").Run() == nil }

func (Apt) InstallCommand(packages []string) []string {
	cmd := []string{"apt-get", "install", "-y"}
	return append(cmd, packages...)
}

// Brew adapts the Homebrew package manager (macOS/Linux).
type Brew struct{}

func (Brew) Name() string      { return "brew" }
func (Brew) NeedsSudo() bool   { return false }
func (Brew) IsAvailable() bool { return exec.Command("brew", "--version").Run() == nil }

func (Brew) InstallCommand(packages []string) []string {
	cmd := []string{"brew", "install"}
	return append(cmd, packages...)
}

// allManagers is the ordered list of supported package managers.
var allManagers = []PackageManager{
	Pacman{},
	Apt{},
	Brew{},
}

// GetPackageManager detects and returns the first available package manager.
// Returns nil if no supported manager is found.
func GetPackageManager() PackageManager {
	for _, m := range allManagers {
		if m.IsAvailable() {
			return m
		}
	}
	return nil
}

// ManagerNames returns sorted names of all supported managers (for display).
func ManagerNames() []string {
	names := make([]string, len(allManagers))
	for i, m := range allManagers {
		names[i] = m.Name()
	}
	sort.Strings(names)
	return names
}
